File: C++/13.代理模式/ProxyPatternDemo.cpp
/**
 * @file ProxyPatternDemo.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-11-12
 *
 * @copyright Copyright (c) 2019
 *
 */

#include "ProxyImage.h"

int main(int argc, char const *argv[]) {
    std::shared_ptr<Image> image = std::make_shared<ProxyImage>("test_10mb.jpg");
    image->Display();

    std::cout << std::endl;

    image->Display();

    return 0;
}

File: Objective-C/13.ProxyPattern/ProxyPatternDemo.m
//
//  ProxyPatternDemo.m
//  13.ProxyPattern
//
//  Created by Layne on 2020/11/3.
//

#import "ProxyImage.h"

int main(int argc, const char* argv[]) {
    @autoreleasepool {
        id<ImageProtocol> image = [[ProxyImage alloc] initWithFileName:@"test_10mb.jpg"];
        // 图像将从磁盘加载
        [image display];
        // 图像不需要从磁盘加载
        [image display];
    }
    return 0;
}
